src/plugins/analysis/device_tree/internal/schema.py

```diff
@@ -0,0 +1,172 @@
+import pydantic
+from pydantic import Field
+from typing import Optional, List, ClassVar
+from tempfile import NamedTemporaryFile
+import pathlib as pl
+
+from .device_tree_utils import StructureBlock, convert_device_tree_to_str, get_model_or_description
+
+
+class IllegalDeviceTreeError(ValueError):
+    pass
+
+
+class IllegalHeaderError(IllegalDeviceTreeError):
+    pass
+
+
+class DeviceTree(pydantic.BaseModel):
+    class Header(pydantic.BaseModel):
+        """The devicetree header as described in [1].
+
+        [1]: https://devicetree-specification.readthedocs.io/en/stable/flattened-format.html#header
+        """
+
+        SIZE: ClassVar[int] = 40
+        MAX_VERSION: ClassVar[int] = 20
+        MAGIC: ClassVar[bytes] = bytes.fromhex('D00DFEED')
+
+        magic: int
+        totalsize: int
+        off_dt_struct: int
+        off_dt_strings: int
+        off_mem_rsvmap: int
+        version: int
+        last_comp_version: int
+        boot_cpuid_phys: int
+        size_dt_strings: int
+        size_dt_struct: int
+
+        @classmethod
+        def from_binary(cls, binary: bytes):
+            """Given the whole device tree binary parses the header and does some sanity checks."""
+            assert len(binary) >= cls.SIZE, f'Given header has size {len(binary)} but it should be at least {cls.SIZE}'
+            header = cls(
+                magic=int.from_bytes(
+                    binary[0:4],
+                    byteorder='big',
+                ),
+                totalsize=int.from_bytes(
+                    binary[4:8],
+                    byteorder='big',
+                ),
+                off_dt_struct=int.from_bytes(
+                    binary[8:12],
+                    byteorder='big',
+                ),
+                off_dt_strings=int.from_bytes(
+                    binary[12:16],
+                    byteorder='big',
+                ),
+                off_mem_rsvmap=int.from_bytes(
+                    binary[16:20],
+                    byteorder='big',
+                ),
+                version=int.from_bytes(
+                    binary[20:24],
+                    byteorder='big',
+                ),
+                last_comp_version=int.from_bytes(
+                    binary[24:28],
+                    byteorder='big',
+                ),
+                boot_cpuid_phys=int.from_bytes(
+                    binary[28:32],
+                    byteorder='big',
+                ),
+                size_dt_strings=int.from_bytes(
+                    binary[32:36],
+                    byteorder='big',
+                ),
+                size_dt_struct=int.from_bytes(
+                    binary[36:40],
+                    byteorder='big',
+                ),
```

Comment:
adding a small helper function could improve readability and reduce code duplication here a lot since this only seems to depend on the offset (byteorder and lenght seem to be constant), i.e. 
```python
def _read_int_from_header(header: bytes, offset: int) -> int:
    return int.from_bytes(header[offset:offset + 4], byteorder='big')
```